File: internal/report/markdown.go
package report

import (
	"fmt"
	"sort"
	"strings"
)

const customCategoryPrefix = "custom:"

// Marker delimits the tool's comment so subsequent runs can find and edit
// it instead of posting a new comment on every push.
const Marker = "<!-- tf-predeploy-firewall:report -->"

var severityEmoji = map[Severity]string{
	SeverityLow:      "🔵",
	SeverityMedium:   "🟡",
	SeverityHigh:     "🟠",
	SeverityCritical: "🔴",
}

var categoryLabel = map[Category]string{
	CategoryUnknownAttribute: "Unknown/hallucinated attribute",
	CategoryUnpinnedVersion:  "Unpinned module or provider version",
	CategoryTutorialPattern:  "Tutorial-copy pattern",
	CategoryForceNewChange:   "ForceNew change on stateful resource",
	CategoryMissingLifecycle: "Missing prevent_destroy",
	CategoryConfirmedReplace: "Confirmed destroy/replace (from terraform plan)",
	CategoryUnexpectedDrift:  "Unexpected drift (from terraform plan)",
	CategoryLargeBlastRadius: "Large blast radius (from terraform plan)",
	CategoryCostImpact:       "Estimated cost impact",
}

// RenderMarkdown builds the full PR comment body for a set of findings.
// blocked indicates whether the configured severity threshold was breached
// by the ACTIVE (non-waived) findings — a waived finding never contributes
// to blocked, but still appears in its own section below so accepting one
// is never silent.
func RenderMarkdown(findings []Finding, threshold Severity, blocked bool) string {
	var b strings.Builder
	b.WriteString(Marker + "\n")
	b.WriteString("## TF Pre-Deploy Firewall\n\n")

	var active, waived []Finding
	for _, f := range findings {
		if f.Waived {
			waived = append(waived, f)
		} else {
			active = append(active, f)
		}
	}

	if len(active) == 0 && len(waived) == 0 {
		b.WriteString("No risk patterns detected in the changed Terraform files. ✅\n")
		return b.String()
	}

	sort.Slice(active, func(i, j int) bool {
		if active[i].File != active[j].File {
			return active[i].File < active[j].File
		}
		return active[i].Line < active[j].Line
	})

	if len(active) == 0 {
		fmt.Fprintf(&b, "✅ No blocking findings — %d finding(s) previously accepted (see below).\n\n", len(waived))
	} else if blocked {
		fmt.Fprintf(&b, "🚫 **Merge blocked** — findings at or above `%s` severity (threshold: `%s`).\n\n", highestSeverity(active), threshold)
	} else {
		fmt.Fprintf(&b, "⚠️ %d finding(s), none reach the `%s` blocking threshold.\n\n", len(active), threshold)
	}

	if len(active) > 0 {
		b.WriteString("| Severity | File | Line | Category | Resource | Detail |\n")
		b.WriteString("|---|---|---|---|---|---|\n")
		for _, f := range active {
			fmt.Fprintf(&b, "| %s %s | `%s` | %d | %s | %s | %s |\n",
				severityEmoji[f.Severity], f.Severity, f.File, f.Line, categoryDisplay(f.Category), resourceCell(f), f.Message)
		}
	}

	renderSuggestions(&b, active)
	renderWaivers(&b, waived)

	return b.String()
}

// renderWaivers lists findings an admin accepted via the control plane
// (Starter+, GET /v1/waivers) — visible so a waiver is a documented
// decision, not a silent removal from the report.
func renderWaivers(b *strings.Builder, waived []Finding) {
	if len(waived) == 0 {
		return
	}
	sort.Slice(waived, func(i, j int) bool {
		if waived[i].File != waived[j].File {
			return waived[i].File < waived[j].File
		}
		return waived[i].Line < waived[j].Line
	})

	fmt.Fprintf(b, "\n<details><summary>%d accepted finding(s) — excluded from the block decision</summary>\n\n", len(waived))
	b.WriteString("| Severity | File | Line | Category | Resource | Detail | Accepted because |\n")
	b.WriteString("|---|---|---|---|---|---|---|\n")
	for _, f := range waived {
		fmt.Fprintf(b, "| %s %s | `%s` | %d | %s | %s | %s | %s |\n",
			severityEmoji[f.Severity], f.Severity, f.File, f.Line, categoryDisplay(f.Category), resourceCell(f), f.Message, f.WaiverNote)
	}
	b.WriteString("\n</details>\n")
}

// resourceCell renders the resource address, linked to the provider
// documentation for its type when one is known. The link rides on the
// address rather than taking a column of its own: a seventh column would
// cost every row width, on a table that is already wide, to carry the same
// word on every line.
func resourceCell(f Finding) string {
	if f.DocURL == "" {
		return "`" + f.Resource + "`"
	}
	return fmt.Sprintf("[`%s`](%s)", f.Resource, f.DocURL)
}

// renderSuggestions appends a collapsible "Suggested fixes" block per
// finding that has one — pasteable HCL, not a computed patch (this tool
// never has write access to the repo). Kept out of the main table since a
// multi-line code block doesn't fit inside a markdown table cell.
func renderSuggestions(b *strings.Builder, sorted []Finding) {
	var any bool
	for _, f := range sorted {
		if f.Suggestion != "" {
			any = true
			break
		}
	}
	if !any {
		return
	}

	b.WriteString("\n### Suggested fixes\n\n")
	for _, f := range sorted {
		if f.Suggestion == "" {
			continue
		}
		fmt.Fprintf(b, "<details><summary><code>%s</code> (%s:%d)</summary>\n\n```hcl\n%s\n```\n\n</details>\n\n",
			f.Resource, f.File, f.Line, f.Suggestion)
	}
}

// categoryDisplay labels a category for the PR comment. Custom rules
// (category "custom:<rule-id>", from internal/customrules) aren't in the
// built-in categoryLabel map — they're rendered as "Custom rule: <id>"
// instead of falling back to an empty string.
func categoryDisplay(c Category) string {
	if label, ok := categoryLabel[c]; ok {
		return label
	}
	if id, isCustom := strings.CutPrefix(string(c), customCategoryPrefix); isCustom {
		return "Custom rule: " + id
	}
	return string(c)
}

func highestSeverity(findings []Finding) Severity {
	highest := SeverityLow
	for _, f := range findings {
		if f.Severity.AtLeast(highest) {
			highest = f.Severity
		}
	}
	return highest
}

File: src/tfpdf/report/markdown.py
"""Le commentaire de pull request.

Port de internal/report/markdown.go.
"""

from __future__ import annotations

from .finding import Finding, Severity
from .ruledocs import category_display

#: Délimite le commentaire de l'outil pour que les exécutions suivantes le
#: retrouvent et le modifient, au lieu d'en poster un nouveau à chaque push.
MARKER = "<!-- tf-predeploy-firewall:report -->"

SEVERITY_EMOJI = {
    Severity.LOW: "🔵",
    Severity.MEDIUM: "🟡",
    Severity.HIGH: "🟠",
    Severity.CRITICAL: "🔴",
}


def _by_file_then_line(finding: Finding) -> tuple[str, int, str, str]:
    """Un ordre **total**, contrairement à celui de Go."""
    return (finding.file, finding.line, str(finding.category), finding.message)


def render_markdown(findings: list[Finding], threshold: Severity | str, blocked: bool) -> str:
    """Construit le corps complet du commentaire de PR pour un ensemble de découvertes."""
    sections: list[str] = [MARKER + "\n", "## TF Pre-Deploy Firewall\n\n"]

    active = [finding for finding in findings if not finding.waived]
    waived = [finding for finding in findings if finding.waived]

    if not active and not waived:
        sections.append("No risk patterns detected in the changed Terraform files. ✅\n")
        return "".join(sections)

    # Clé totale, contrairement au sort.Slice de Go : deux découvertes sur le
    # même fichier et la même ligne sont départagées par la catégorie puis le
    # message, et non par une permutation arbitraire. Voir _by_file_then_line.
    active.sort(key=_by_file_then_line)

    if not active:
        sections.append(
            f"✅ No blocking findings — {len(waived)} finding(s) previously "
            "accepted (see below).\n\n"
        )
    elif blocked:
        sections.append(
            f"🚫 **Merge blocked** — findings at or above `{highest_severity(active)}` "
            f"severity (threshold: `{threshold}`).\n\n"
        )
    else:
        sections.append(
            f"⚠️ {len(active)} finding(s), none reach the `{threshold}` blocking threshold.\n\n"
        )

    if active:
        sections.append("| Severity | File | Line | Category | Resource | Detail |\n")
        sections.append("|---|---|---|---|---|---|\n")
        for finding in active:
            sections.append(
                f"| {SEVERITY_EMOJI.get(finding.severity, '')} {finding.severity} | `{finding.file}` | "
                f"{finding.line} | {category_display(finding.category)} | {resource_cell(finding)} | "
                f"{finding.message} |\n"
            )

    _render_suggestions(sections, active)
    _render_waivers(sections, waived)

    return "".join(sections)


def _render_waivers(sections: list[str], waived: list[Finding]) -> None:
    """Liste les découvertes qu'un administrateur a acceptées via le plan de contrôle (Starter et
    plus, GET /v1/waivers)."""
    if not waived:
        return
    waived = sorted(waived, key=_by_file_then_line)

    sections.append(
        f"\n<details><summary>{len(waived)} accepted finding(s) — excluded from "
        "the block decision</summary>\n\n"
    )
    sections.append(
        "| Severity | File | Line | Category | Resource | Detail | Accepted because |\n"
    )
    sections.append("|---|---|---|---|---|---|---|\n")
    for finding in waived:
        sections.append(
            f"| {SEVERITY_EMOJI.get(finding.severity, '')} {finding.severity} | `{finding.file}` | "
            f"{finding.line} | {category_display(finding.category)} | {resource_cell(finding)} | "
            f"{finding.message} | {finding.waiver_note} |\n"
        )
    sections.append("\n</details>\n")


def resource_cell(f: Finding) -> str:
    """Rend l'adresse de la ressource, liée à la documentation du fournisseur pour son type quand
    elle est connue."""
    if not f.doc_url:
        return "`" + f.resource + "`"
    return f"[`{f.resource}`]({f.doc_url})"


def _render_suggestions(sections: list[str], sorted_findings: list[Finding]) -> None:
    """Ajoute un bloc repliable « Suggested fixes » par découverte qui en a un."""
    if not any(finding.suggestion for finding in sorted_findings):
        return

    sections.append("\n### Suggested fixes\n\n")
    for finding in sorted_findings:
        if not finding.suggestion:
            continue
        sections.append(
            f"<details><summary><code>{finding.resource}</code> ({finding.file}:{finding.line})</summary>"
            f"\n\n```hcl\n{finding.suggestion}\n```\n\n</details>\n\n"
        )


def highest_severity(findings: list[Finding]) -> Severity:
    highest = Severity.LOW
    for finding in findings:
        if finding.severity.at_least(highest):
            highest = finding.severity
    return highest
